### PR TITLE
Rework PR template to improve quality

### PR DESCRIPTION
**Pull request by jvanderhoof**
This change adds a list of checks tasks an engineers should have performed before opening a PR and requesting a code review.

The goal is improve the quality of code as well as ease PR reviews.

`cyberark/conjur` is intended to be the first repository for this change.  Assuming it's accepted, we'll roll this out to the other repositories.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] All debugging code has been removed.
+- [ ] The code is as modular as possible.
+- [ ] All code follows the defined architecture.
+- [ ] The feature implementation follows the Feature Spec.
+- [ ] Good names are used for classes, methods, and variables.
+- [ ] Code Climate linting issues have been resolved.
 
-> Has this change been documented (Readme, docs, etc.)?
+### Logging
+- [ ] Logging is available.
```
Comment (garkler) [inline .github/PULL_REQUEST_TEMPLATE.md]: Can we add something about all error messages being clear about what the error message is ?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] A review can step through the PR commit by commit to gain a complete understanding of the functionality.
+
+### Code Documentation
+- [ ] All edge-case handling is described.
+- [ ] The use and function of any new third-party libraries is documented.
+- [ ] The Version and Changelog has been updated.
+
+### Functional Documentation
+- [ ] The Readme has been updated (if relevant).
+- [ ] A PR is opened for any documentation changes (or, if large, a story has been created to capture these document changes).
```
Comment (garkler) [inline .github/PULL_REQUEST_TEMPLATE.md]: In terms of supportability, I would like to have the Feature Doc start capturing important messages \ error messages so we can create a troubleshooting guide for new features.
I don't know if its too advanced at this point ? But throwing it out there
@alexkalish , @jvanderhoof wdyt ?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -1,16 +1,40 @@
...
-#### Screenshots (if appropriate)
-#### Has the Version and Changelog been updated?
-#### Questions:
-> Does this work have automated integration and unit tests?
 
-> Can we make a blog post, video, or animated GIF of this?
+## Checklist
+### Code
+- [ ] The code works.
+- [ ] The code performs its intended function and the logic is correct.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]: Would this cover `The code works` as well? Or is there a distinction between the two that should be called out?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] The code easy to understand.
+- [ ] Redundant or duplicate code has been removed.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Redundant, duplicate, and commented code has been removed.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] All debugging code has been removed.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] All debugging or commented out code has been removed.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] The code is as modular as possible.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]: This is fine, but I know for myself, it would be helpful to perhaps also point to a design guide that helps elaborate what it means to be modular. For example:

"The code adheres to the [SOLID principles](https://en.wikipedia.org/wiki/SOLID)."

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] All code follows the defined architecture.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]: The template should probably have a place to link to the feature spec so that it's easy for the reviewer to find the architecture and requirements the PR is supposed to fulfill.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] The feature implementation follows the Feature Spec.
+- [ ] Good names are used for classes, methods, and variables.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]: Same thing as the modular above, without some guidance, this is a fairly arbitrary checkbox. This is one resource that provides some helpful guidelines: https://medium.com/coding-skills/clean-code-101-meaningful-names-and-functions-bf450456d90c

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -1,16 +1,40 @@
...
-> Has this change been documented (Readme, docs, etc.)?
+### Logging
+- [ ] Logging is available.
+- [ ] All logging code provides the correct visibility for the appropriate logging level (ex. DEBUG is verbose, FATAL shows only catastrophic events)
+- [ ] No credentials or secrets are visible through logs.
 
-> Does the knowledge base need an update?
+### Testing
+- [ ] The code is testable.
+- [ ] The code does not add any hidden dependencies.
```
Comment (micahlee) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'm not sure what this means.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -1,16 +1,40 @@
...
+## Checklist
+### Code
+- [ ] The code works.
+- [ ] The code performs its intended function and the logic is correct.
+- [ ] The code easy to understand.
+- [ ] Redundant or duplicate code has been removed.
+- [ ] All debugging code has been removed.
+- [ ] The code is as modular as possible.
+- [ ] All code follows the defined architecture.
+- [ ] The feature implementation follows the Feature Spec.
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]: The new document name is `Technical Design and Delivery`.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] Good names are used for classes, methods, and variables.
+- [ ] Code Climate linting issues have been resolved.
 
-> Has this change been documented (Readme, docs, etc.)?
+### Logging
+- [ ] Logging is available.
+- [ ] All logging code provides the correct visibility for the appropriate logging level (ex. DEBUG is verbose, FATAL shows only catastrophic events)
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]: I was actually just looking at Oren's logging PRs and the docs that Sigal made.  Similar to what @micahlee said above, I think some guidance would be beneficial outside of the checklist.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -1,16 +1,40 @@
...
+- [ ] No credentials or secrets are visible through logs.
 
-> Does the knowledge base need an update?
+### Testing
+- [ ] The code is testable.
+- [ ] The code does not add any hidden dependencies.
+- [ ] Tests exist and are they comprehensive.
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] Tests exist and they are comprehensive.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -1,16 +1,40 @@
...
-#### Has the Version and Changelog been updated?
-#### Questions:
-> Does this work have automated integration and unit tests?
 
-> Can we make a blog post, video, or animated GIF of this?
+## Checklist
+### Code
+- [ ] The code works.
+- [ ] The code performs its intended function and the logic is correct.
+- [ ] The code easy to understand.
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] The code is easy to understand.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -1,16 +1,40 @@
...
+### Testing
+- [ ] The code is testable.
+- [ ] The code does not add any hidden dependencies.
+- [ ] Tests exist and are they comprehensive.
+- [ ] Unit tests actually test that the code is performing the intended functionality.
+
+### Commit History
+- [ ] Commit messages accurately describe the changes.
+- [ ] Each commit delivers a discrete set of value.
+- [ ] A review can step through the PR commit by commit to gain a complete understanding of the functionality.
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]: Has there been general agreement that this is how commit history should be structured?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (alexkalish) [reply .github/PULL_REQUEST_TEMPLATE.md]: I think it belongs in the Technical D&D doc.  Feels too low level for the Feature Proposal.  Or we need a need "Supportability" document?  Dunno.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -1,16 +1,40 @@
-#### What does this PR do?
```
Comment (alexkalish) [inline .github/PULL_REQUEST_TEMPLATE.md]: I wouldn't totally ditch all written prose in the template.  I think that there is value in at least a bulleted list of changes/additions, or a few sentences describing the work.  But I agree that the there were too many sections here before.

**Comment by jonahx** [comment]
@jvanderhoof We should discuss the nuances in person today when we talk, but my general feeling is twofold:

Pros:

- It's good to explicit about all these items
- Reminders are good
- Forced reflection is helpful

Cons:

- It's adds a lot of boilerplate
- The GH UI doesn't allow you to "check" checkboxes while in preview mode.
- For most PRs, only a subset of the items will be relevant

So I think it's important to have flexibility around deleting items when they don't apply.  If you had to go through each item and type in X or provide an explanation why you didn't, on every PR, it would be too onerous, and would become a nuisance.

Our goal is to improve quality and have a good developer experience with minimal busy work.   To keep this from becoming heavyweight, my suggestions are:

- Add flexibility into the process
- Cut the total number of bullet points down by around half ("If everything is important, nothing is")

For the 2nd point, it's not even that I think any of them are bad.  I just think the outcome would be better by focusing on the what we *really* care about, and forcing ourselves to revise mercilessly.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (rafis3) [reply .github/PULL_REQUEST_TEMPLATE.md]: But I also suggest that we would have a section that requests from the engineer to give a small description on the PR, explain what was the intended functionality and if there is any unchecked box left, justify it. I think that rarely there will be good reasons to leave some of them unchecked.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 13 above)
Comment (rafis3) [inline .github/PULL_REQUEST_TEMPLATE.md]: What do you think about adding:
`The code follows the language standards (e.g. indentions, casing, naming conventions, location of brackets, etc)`

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 12 above)
Comment (rafis3) [reply .github/PULL_REQUEST_TEMPLATE.md]: To follow along with the rest of the places, maybe we should add a guideline on what makes tests comprehensive? @andytinkham what do you think?

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (rafis3) [inline .github/PULL_REQUEST_TEMPLATE.md]: How about adding a section for `Build` and add under it:
`The build completes successfully in Jenkins`

Comment (micahlee) [reply .github/PULL_REQUEST_TEMPLATE.md]: Would this add information for the submitter/reviewer not already provided by the Jenkins "GitHub Check"?

**Comment by garkler** [comment]
Are the updates to the changelog.md done per PR ? If so, can we have this added to the template? (and if not, where can I make sure its done)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 16]
```diff
@@ -1,16 +1,40 @@
...
 Connected to [relevant GitHub issues, eg #76]
-#### Where should the reviewer start?
-#### How should this be manually tested?
-#### Screenshots (if appropriate)
-#### Has the Version and Changelog been updated?
-#### Questions:
-> Does this work have automated integration and unit tests?
 
-> Can we make a blog post, video, or animated GIF of this?
+## Checklist
```
Comment (izgeri) [inline .github/PULL_REQUEST_TEMPLATE.md]: I would suggest we have all of the boxes checked at first, and the checklist instructions should ask the PR submitter to read through the list and uncheck any that don't apply. I agree with @rafis3 below that the vast majority of the time, all of these will be checked (and in practice I doubt if people will actually go through and check > 10 boxes)

Comment (izgeri) [inline .github/PULL_REQUEST_TEMPLATE.md]: I would suggest we have all of the boxes checked at first, and the checklist instructions should ask the PR submitter to read through the list and uncheck any that don't apply. I agree with @rafis3 below that the vast majority of the time, all of these will be checked (and in practice I doubt if people will actually go through and check > 10 boxes)

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 7 above)
Comment (izgeri) [reply .github/PULL_REQUEST_TEMPLATE.md]: On the other hand, if the PR is linked to an issue and that issue is linked to an epic - there is a trail to follow. I'm not sure it's realistic to expect every PR in an epic to link to the feature spec, and it's likely redundant if everything else has been linked correctly.

**Comment by ryanprior** [comment]
tl;dr: this is a tire fire, scrap it. Thoughts in detail:

1. Who's trying to fill out a 25-part form every time they submit a PR? Who's
   trying to check the veracity of 25 form fields every time they do a code
   review?
   
   The most common thing people did with the old 9-part form was to delete a few
   of the parts, because many PRs didn't really need all that. Who benefits from
   upping the initial commitment to 25 form-parts?

   The suggestion of checking all the boxes by default does not help this at
   all: it reduces the amount of work it takes to lazily say "yup everything's
   good" but if you're really considering each point, it takes the same amount
   of time whether the boxes start checked or not. Checking or un-checking the
   boxes is the very quickest and most trivial part any of this.
   
   Comically, a bunch of the comments here are "can we also add [yet another
   section]?" as if the perception is that PRs would be better if people made 30
   certifications instead of 25.
   
2. Where's the evidence showing that more time spent with this type of
   subjective checklist (more on this soon) improves software quality?
   Incidentally, if this PR actually used the old template, it might have
   discussed the supporting context, making it easier to review without such
   guesswork.
   
   We have plenty of evidence that checklists improve patient outcomes in health
   care settings, if they are the right sort of checklist. For example, the
   American Hospital Association describes that "to prevent process breakdowns
   due to human factors, each checklist identifies the top 10 evidence-based
   interventions that health care organizations can implement and test to reduce
   harm." ([source](http://www.hpoe.org/Reports-HPOE/CkLists_PatientSafety.pdf))
   
   Let's break that down.
   *  They limit checklists to 10 most urgent items.
   *  Each item must be an intervention: it needs to be actionable, such that
      when you read a checkbox you know right away whether you can check it or
      not, and if you are unable to check it, there's a clear intervention to
      take in order to check off.
   *  Each intervention is itself evidence-based. No intervention is included
      based on guesswork. The checklist is a top-ten list of the highest-impact
      proven interventions to increase safety and support patient outcomes.
      
   Compare to this checklist: 25 items and counting, many of which do not
   prescribe any clear intervention, and none of which are accompanied with
   specific definitions of terms and evidence that the intervention is
   effective.

3. How is this template remotely appropriate for anybody outside of CyberArk
   engineering? I count eleven parts of this which are highly subjective,
   basically judgment calls. Somebody who's doing Conjur stand-ups every week
   and getting active mentoring from Conjur core contributors can maybe make
   those judgment calls, but how's somebody outside that culture supposed to
   represent whether their PR is "easy to understand" and "as modular as
   possible" etc etc ad nauseum?
   
   A PR template is shown to anybody who opens a PR, no matter their background.
   The text of this is effectively a ward against collaboration. Worse yet, none
   of these things are hyperlinked to allow a less experienced person to
   understand what's being asked. Supposing I don't know where to find the
   standard for "good names for classes, methods, and variables," why would I
   feel confident certifying that my PR meets this standard?
   
   This would be one thing if it were an internal JIRA workflow, but as a
   community-facing PR template, it's hostile, overbearing, and tedious.
   
4. A number of these are essentially asking the PR author to certify that they
   haven't made any mistakes. Where's the evidence that people can be relied
   upon to tell you whether they've made any mistakes in the first place?
   Where's the value in asking that versus building tools to detect those
   mistakes or a guide to teach people how to spot those mistakes?
   
   Specifically guilty of this:
   *  "The code works." As others have noted in the comments here, if it doesn't
      work, I screwed up & probably didn't mean to. The proof of this is in
      testing, so why make me certify that I didn't screw up?
   *  "No credentials or secrets are visible through logs." What is this
      intended to catch? Somebody who intentionally prints secrets to stdout in
      their PR but didn't realize that was undesirable until they opened a PR?
   *  "The code does not add any hidden dependencies." Beyond this being
      somewhat subjective as per (2), is there a way to test this or a guide to
      help people check? If a PR author doesn't realize they're adding a hidden
      dependency, how does it help to have them check this box?

**Comment by rafis3** [comment]
Hey @ryanprior. Thanks for the feedback! :)
In my mind, this checklist was meant to help the contributor think about her added feature.

If a certain checkbox is irrelevant, it can be removed. But the process of getting to the conclusion that it's not relevant was done with a conscious decision, after thinking it through.

If a checkbox is relevant but the requirement is not met, it's a great point for discussion between the contributor and the reviewer. It serves as a "heads up" for the reviewer.

If a checkbox was checked but the requirement hasn't been actually met, that happens. I come from the assumption that the contributor is trying to do the best possible work she can and we all make mistakes. The reviewer can still check and validate to make sure the change is done as expected.

And I definitely agree that we should add more links to pages that explain about the expected criteria, to create a unified language, a set of expectations, and, as you said, to allow the contributor to learn from it. I think that this list is especially helpful for unexperienced contributors and can help them to make sure their feature is not missing any part of the product (such as logging, documentation, audit, code quality and so on).

I think that the advantage of that checklist being in the PR, is that these questions come after you worked on the feature, thought it through, decided it's good and should be merged. Then comes this checklist, that can help you verify you haven't missed anything. 

Does that make sense? What do you think?

**Comment by jvanderhoof** [comment]
I'm closing this as I have not had time to focus on this effort.  I'll re-open in the future.
